text_generation/causal_lm/cpp/prompt_lookup_decoding.cpp

```diff
@@ -0,0 +1,308 @@
+        size_t K = candidates.size();
+
+        // K + 1 tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, K + 1});
+        input_ids.data<int64_t>()[0] = first_token;
+        for (int i = 0; i < K; i++) {
+            input_ids.data<int64_t>()[i + 1] = candidates[i];
+        }
+
+        attention_mask.set_shape({BATCH_SIZE, seq_len + K + 1});
+        std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
+
+        position_ids.set_shape({BATCH_SIZE, K + 1});
+        std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), seq_len);
+
+        model.infer();
+
+        data_logits = logits.data<float>();  // [BATCH_SIZE, 1 + K, vocab_size]
+
+        // Find candidates matches.
+        // Accept first token (if not EOS) as it generated from the valid input_ids sequence.
+        // Iterate only if canidate token matches model predicted token.
+        // Accept last token as it generated from the valid predicted tokens sequence.
+        size_t accepted_tokens_number = 0;
+        for (size_t i = 0; i < K + 1; i++) {
+            auto start = data_logits + vocab_size * i;
+            auto stop = data_logits + vocab_size * (i + 1);
+            auto candidate_out_token = std::max_element(start, stop) - start;
+            out_token = candidate_out_token;
+
+            if (candidate_out_token == EOS_TOKEN) {
+                break;
+            }
+
+            text_streamer.put(candidate_out_token);
+            full_input_ids.push_back(candidate_out_token);
+            accepted_tokens_number++;
+
+            if (i == K || candidate_out_token != candidates[i] || seq_len + i >= max_sequence_length) {
+                break;
+            }
+        }
+
+        if (accepted_tokens_number > 0) {
+            candidateGenerator.update_cadidate_strategy(accepted_tokens_number - 1);
+        }
+
+        // After the inference request, key/values have shape [BATCH_SIZE, seq_len + K, vocab_size].
+        // Increment the sequence length by the number of matched tokens, and
+        // trim the KV cache to match the new sequence length.
+        seq_len += accepted_tokens_number;
+        update_kv_cache(model, SEQ_LEN_AXIS, seq_len);
+
+        first_token = out_token;
+    }
+
+    text_streamer.end();
+    // Model is stateful which means that context (kv-cache) which belongs to a particular
+    // text sequence is accumulated inside the model during the generation loop above.
+    // This context should be reset before processing the next text sequence.
+    // While it is not required to reset context in this sample as only one sequence is processed,
+    // it is called for education purposes:
+    model.reset_state();
+} catch (const std::exception& error) {
+    std::cerr << error.what() << '\n';
+    return EXIT_FAILURE;
+} catch (...) {
+    std::cerr << "Non-exception object thrown\n";
+    return EXIT_FAILURE;
+}
```

Comment:
```suggestion
}

```

Reply:
I was relying on clang-format config from OpenVINO which doesn't force new line EOF. Tried to apply ['InsertNewlineAtEOF'](https://releases.llvm.org/16.0.0/tools/clang/docs/ClangFormatStyleOptions.html#insertnewlineateof) but it's available for clang-format 16 only when OV target version as I understand 9.